### PR TITLE
feat(nm): add configuration enforcement

### PR DESCRIPTION
**Pull request by mattdibi**
**Related Issue:** https://github.com/eclipse/kura/pull/4422

By “_Configuration enforcement_” we mean the ability to roll-back the system network configuration to the one set by Kura when a user changes it manually, so that the Kura snapshot and the system network configuration are consistent.

This PR monitors NetworkManager DBus signals related to configuration changes and performs a configuration roll-back in case an unauthorized change is detected.

**Note**: we'll monitor only NM changes that happens through NetworkManager. If a user changes the network configuration without using neither Kura nor NetworkManager we won't catch the change.

---

Implementation: we monitor the `CONFIG`and `DISCONNECTED` events. If they're captured a roll-back is issued. We disarm the monitor of the event when Kura starts configuring the interfaces.

**Note**: with this implementation the `DeviceAdded` signal handler is rendered unnecessary because, when a new device is plugged in, it triggers the same signals that we monitor for the configuration enforcement.

### CODE REVIEW

**Comment by mattdibi** [comment]
@pierantoniomerlino @salvatore-coppola I went heavy on the unit tests side so that it's clear the assumption we're making for the configuration enforcement to work.

The thing I don't like about all of this is how I simulated the DBus signal interception:

https://github.com/eclipse/kura/blob/4f7eae1bb951a9bfe5b38ece9c9792ebb1be2d57/kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/NMDbusConnectorTest.java#L805-L821

It's a dirty hack but, being the DBus connection a mock, I wasn't able to trigger the signal interception in any other way. If I set the mock as "CALLS_REAL_METHODS" it breaks because there's no DBus during testing. Tell me what you guys think.

**Comment by mattdibi** [comment]
### Test performed

#### Environment

Raspberry Pi 4 running Raspbian OS NetworkManager 1.30.6

```bash
uname -a
Linux raspberrypi 5.15.61-v8+ #1579 SMP PREEMPT Fri Aug 26 11:16:44 BST 2022 aarch64 GNU/Linux
```

#### Tests

##### Test 1: Pluggin in a new device

Plugging in a new USB to Ethernet adapter results in NM trying to configuring it. This will send a `CONFIG` state signal on the DBus which gets intercepted:

```
2023-03-08T08:57:32,312 [DBus Worker Thread-3] INFO  o.e.k.n.NMConfigurationEnforcementHandler - Network change detected on interface /org/freedesktop/NetworkManager/Devices/7. Roll-back to cached configuration
```

No configuration roll-back is triggered on disconnecting the adapter.

##### Test 2: Disconnection through `nmcli`

```bash
sudo nmcli device disconnect wlan0
```

Results in a roll-back

```
2023-03-08T08:58:40,397 [DBus Worker Thread-2] INFO  o.e.k.n.NMConfigurationEnforcementHandler - Network change detected on interface /org/freedesktop/NetworkManager/Devices/4. Roll-back to cached configuration
```

##### Test 3: Configuration update through `nmcli`

```bash
pi@raspberrypi:~ $ sudo nmcli dev modify wlan0 ipv4.address 172.16.1.4
Connection successfully reapplied to device 'wlan0'.
pi@raspberrypi:~ $ nmcli con show
NAME                   UUID                                  TYPE      DEVICE
kura-eth0-connection   0c70961a-dccf-4015-a8a3-115ab4ca284e  ethernet  eth0
kura-wlan0-connection  6eeaf256-2ee9-49eb-81db-d8f82ee8b2d8  wifi      wlan0
pi@raspberrypi:~ $ sudo nmcli con up 6eeaf256-2ee9-49eb-81db-d8f82ee8b2d8
Error: Connection activation failed: The base network connection was interrupted
Hint: use 'journalctl -xe NM_CONNECTION=6eeaf256-2ee9-49eb-81db-d8f82ee8b2d8 + NM_DEVICE=wlan0' to get more details.
```

```
2023-03-08T09:01:10,224 [DBus Worker Thread-3] INFO  o.e.k.n.NMConfigurationEnforcementHandler - Network change detected on interface /org/freedesktop/NetworkManager/Devices/4. Roll-back to cached configuration
```

---

No configuration loop was triggered during testing thanks to #4422

**Comment by salvatore-coppola** [comment]
> @pierantoniomerlino @salvatore-coppola I went heavy on the unit tests side so that it's clear the assumption we're making for the configuration enforcement to work.
> 
> The thing I don't like about all of this is how I simulated the DBus signal interception:
> 
> https://github.com/eclipse/kura/blob/4f7eae1bb951a9bfe5b38ece9c9792ebb1be2d57/kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/NMDbusConnectorTest.java#L805-L821
> 
> It's a dirty hack but, being the DBus connection a mock, I wasn't able to trigger the signal interception in any other way. If I set the mock as "CALLS_REAL_METHODS" it breaks because there's no DBus during testing. Tell me what you guys think.

it's something hard to simulate, in that case it is better to rely on testing on real devices.

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMConfigurationEnforcementHandler.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class NMConfigurationEnforcementHandler implements DBusSigHandler<Device.StateChanged> {
+
+    private static final Logger logger = LoggerFactory.getLogger(NMConfigurationEnforcementHandler.class);
+    private final NMDbusConnector nm;
+
+    public NMConfigurationEnforcementHandler(NMDbusConnector nmDbusConnector) {
+        this.nm = Objects.requireNonNull(nmDbusConnector);
```
Comment (salvatore-coppola) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMConfigurationEnforcementHandler.java:29]: Didn't you suggest me to use the NMDbusConnector.getInstance in place of passing the NMDbusConnector reference ?

Comment (mattdibi) [reply kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMConfigurationEnforcementHandler.java:29]: Yes ... and I was absolutely ✨**wrong**✨ !

I started my implementation like that and then realised how wrong I was when implementing tests. Without dependency injection I was completely unable to mock the behaviour during testing and all tests broke.

Sorry 🙏

**Review by pierantoniomerlino — APPROVED** [review]

**Review by salvatore-coppola — APPROVED** [review]

**Pull request merged**